paddle/phi/backends/gpu/cuda/cuda_graph.h

```diff
@@ -88,18 +87,101 @@ class CUDAGraphContextManager {
 
 class CUDAKernelParams {
  public:
-  explicit CUDAKernelParams(const cudaKernelNodeParams *params)
-      : params_(params) {}
-
-  const void *func() const { return params_->func; }
+  explicit CUDAKernelParams(void **params) : kernelParams(params) {}
 
   template <typename T>
   T &As(size_t idx) const {
-    return *reinterpret_cast<T *>(params_->kernelParams[idx]);
+    return *reinterpret_cast<T *>(kernelParams[idx]);
+  }
+
+  void **getParams() const { return kernelParams; }
+
+ private:
+  void **kernelParams;
+};
+
+using cudaGraphExecuterSetter_t = std::function<void(cudaGraphExec_t)>;
+
+//  This class offers a interface for launching CUDA kernels in CUDA Graph, we
+//  utilize the `cudaGraphExecKernelNodeSetParams` function for parameter setup.
+//  Launching kernels via this class ensures proper management.
+//
+//  NOTE: It's essential that the first parameter for any kernel launched
+//  through this class is an `unsigned int` identifier. This identifier plays a
+//  crucial role in linking the CUDA kernel to its corresponding CUDA graph
+//  node. We tag each kernel launch with a unique identifier to maintain
+//  structured linkage with its CUDA graph node.
+//
+//  NOTE: This class use a singleton design pattern ensures there's only a
+//  single global instance accessible via the `Instance()` method.
+//
+//  === Callback Definitions ===
+//
+//  [Parameter Setter Callback]
+//  Sets the kernel's parameters BEFORE activating the CUDA graph. It enables
+//  dynamic determination and setup of kernel arguments.
+//
+//  parameterSetter_t parameterSetter = [saved_state](CUDAKernelParams &param){
+//      // Code to compute and the parameter values from the saved_state
+//      // ...
+//      param.As<type>(idx) = calculated_value;
+//  };
+//
+//  [CUDA Kernel Callback]
+//  Acts as the launcher for the kernel. It accepts an `unsigned int` identifier
+//  and uses it for the kernel launch.
+//
+//  cudaKernelCallback_t cudaKernelCallback = [=](unsigned int id) {
+//      kernel<<<>>>(id, ...);  // Launching the kernel with id
+//  };
+//
+//  [Retrieving CUDA Function]
```

Comment:
The API is intended to be invoked within each library, such as user-defined operators. Consequently, users of CUDAGraphNodeLauncher are accountable for obtaining the cudaFunction_t (CUFunction) structure, and this API should not be directly called by us (in the CUDAGraphNodeLauncher class).